docs/proposal-007-apply.md

```diff
@@ -0,0 +1,137 @@
+# `eksctl apply`
+
+This proposal addresses a GitOps-compatible subcommand of `eksctl` called
+_tentatively_ `apply`. Its purpose is to subsume the various imperative commands
+`eksctl` provides now into one `apply` command that reconciles the actual state
+of an EKS cluster with the intended state as specified in the `eksctl` config file.
+
+This proposal is only concerned with the `apply` subcommand. Existing behavior
+and commands are unaffected. Continuing to support them shouldn't complicate
+the implementation too much as they are more or less subsets of
+operations performed by `apply`.
+
+## Questions to be answered
+
+-   Does the config represent the entire state of the config vs a partial
+    spec?
+    -   e.g. if I `apply` with a list of 1 nodegroups, are any other existing
+        nodegroups deleted?
+    -   do we also delete non-eksctl managed/created resources? i.e. does `eksctl`
+        own the cluster by default? (potential for configurability here)
+-   Surfacing which fields are immutable (i.e. only settable during
+    cluster/nodegroup creation)
+    -   One possibility might be structuring the config in such a way that
+        immutable fields are separated out
+-   How to introduce gradual, partial support?
+-   How do we handle the config
+    1. Leave config alone
+    2. Introduce a _new_ type which only specifies
+       those parts of the config currently supported by `apply`
+       and put it under a new top level field
+    3. Restructure the config gradually
+
+## Proposal
+
+1. `apply` is put behind the explicit "experimental" flag (cf `enable profile/repo`)
+2. We take full ownership of `eksctl`-managed resources (i.e. delete missing resources by default)
+3. Options for non-`eksctl`-managed resources as well as resources where
+   ownership is unknowable are covered below
+4. Immutability is surfaced by erroring when discrepancies are detected
+5. We gradually expand `apply` support to different parts of the cluster
+6. As we expand `apply` support, we reevaluate the config structure
+
+Some further discussion/justification follows.
+
+### Experimental flag
+
+We make it explicit that `apply` is experimental, very open to feedback and
+subject to change.
+
+### Ownership
+
+Ownership comes into play most importantly with deletion.
+
+#### `eksctl`-managed resources
+
+Given the config:
+
+```
+nodeGroups:
+  - name: ng-2
+    desiredCapacity: 1
+```
+
+we discover an existing nodegroup `ng-1` with the `eksctl.io/owned: true` tag (for example),
+do we delete it?
+
+The proposed answer is yes, it should be deleted. With `apply` we assume that all
+previous changes initiated by `eksctl` were initiated through the same config
+file/reconciliation process and that this nodegroup must have been
+deleted from this file.
```

Comment:
I think while the imperative commands are so easy and feature rich (you have flags for doing a lot of what you can do in the config file) this is probably going to lead to a lot of users accidentally deleting things because they made it imperatively. 

I think we can help reduce this by actively driving users towards using apply and away from the imperative commands by:
- Reducing the functionality of imperative commands. It reminds me of how kubectl's imperative commands are great for beginners and getting introduced to kubectl, but the moment you need to do anything more complex you are forced into a config file, which is great
- Being able to generate a config file for existing imperatively created resources. This might not be feasible, but for getting folks onboard to apply with existing clusters having a `eksctl get config` or something to that affect would be  great

I know you called out in the intro `Existing behavior and commands are unaffected`, but I couldn't help myself 😁

_Originally posted by @aclevername_

Reply:
This could be helped by not supporting imperative commands for a new config. Like once you specify nodegroups in v1alpha6 or whatever, you can no longer run create nodegroup. Or if we have a new kind for reconciliation.

I don't think removing functionality would go over well with anyone

Reply:
> This could be helped by not supporting imperative commands for a new config

+1 this

Reply:
> This could be helped by not supporting imperative commands for a new config

sounds good 💯 

> I don't think removing functionality would go over well with anyone

I'm definitely biased here, but using `apply` and config file to me is without a doubt the best way to use eksctl and something that we as a project should push users towards (the later might be a controversial opinion). "pushing" users towards that feels like something worth considering. A greater adoption of `apply` would also most likely also help get users into a space where using GitOps is more applicable and perhaps open a pathway for them using more of WW's tools. Reducing the imperative commands functionality is definitely a big thing and not something we should do lightly, but I don't think it should be dismissed.

Looking at it from an alternative point of view, if we view `v1.0` as a fresh start for eksctl, what would you want it to look like? I'm masively in favour of how kubectl handles this. It has barebone imperative commands for getting beginners active quickly and quickly pushes you towards config files for serious usage.

The above perhaps belongs in a different discussion, but I wanted to make sure my point made sense 😄

Reply:
Completely agree. v1 let's us do away with all imperative commands in a socially acceptable way.

Reply:
i don't think getting rid of imperative commands has much to do with this proposal/`eksctl apply`